From d3e5dc9515779d74a24d144e2343265fcac574b4 Mon Sep 17 00:00:00 2001
From: Lily Li <lali@codeaurora.org>
Date: Tue, 22 Oct 2019 16:57:51 +0800
Subject: [PATCH 26/42] omxvideoenc: dynamic frame rate setting support for omx
 video encoder.

Set new framerate through OMX_SetConfig, when format changed while omx
encoder in Executing.

cherry-pick from 1806777

Change-Id: I95d0c640120b4c3ea38e571b475ea40b210bfe83
Signed-off-by: Lily Li <lali@codeaurora.org>
---
 omx/gstomxvideoenc.c | 29 +++++++++++++++++++++++++++++
 1 file changed, 29 insertions(+)

diff --git a/omx/gstomxvideoenc.c b/omx/gstomxvideoenc.c
index 90d370d..42f5d30 100644
--- a/omx/gstomxvideoenc.c
+++ b/omx/gstomxvideoenc.c
@@ -2241,6 +2241,35 @@ gst_omx_video_enc_set_format (GstVideoEncoder * encoder,
   GST_DEBUG_OBJECT (self, "Setting new format %s",
       gst_video_format_to_string (info->finfo->format));
 
+  // If framerate changed during Executing state set new framerate through
+  // OMX_SetConfig
+  if (gst_omx_component_get_state (self->enc,
+      GST_CLOCK_TIME_NONE) == OMX_StateExecuting && self->input_state) {
+    GstVideoCodecState *prevState = self->input_state;
+    GstVideoInfo *prevInfo = &prevState->info;
+
+    if (prevInfo->fps_n != info->fps_n ||
+            prevInfo->fps_d != info->fps_d) {
+      GST_DEBUG_OBJECT (self, "Frame rate changed from %u to %u",
+              prevInfo->fps_n, info->fps_n);
+      OMX_CONFIG_FRAMERATETYPE enc_framerate;
+      OMX_ERRORTYPE err;
+
+      gst_omx_video_enc_flush (encoder);
+
+      GST_OMX_INIT_STRUCT (&enc_framerate);
+      enc_framerate.nPortIndex = self->enc_out_port->index;
+      FractionToQ16(enc_framerate.xEncodeFramerate, (int) (info->fps_n * 2), 2);
+      err =
+         gst_omx_component_set_config (self->enc,
+         OMX_IndexConfigVideoFramerate, &enc_framerate);
+      GST_DEBUG_OBJECT (self, " Setting Video fps with err: %s (0x%08x)",
+          gst_omx_error_to_string (err), err);
+
+      return TRUE;
+    }
+  }
+
   gst_omx_port_get_port_definition (self->enc_in_port, &port_def);
 
   needs_disable =
-- 
2.7.4

